Destination container in the linkerd-destination pod panics when using deployments with headless services

### What is the issue?

When using the `edge-24.3.4` version (Same problem currently exists in main) and you enable linkerd (`linkerd.io/inject: enabled`) in deployments that use headless services under certain conditions (Probably the nature of the application deployed matters as we are not suffering this bug in all the cases), the linkerd destination container in the linkerd destination pod panics after establishing a watch on a workload with the following error (Debug log level enabled):

```bash
{"addr":":8086","component":"server","context-ns":"example-namespace","context-pod":"example-pod-6b686c4df-fh4km","level":"debug","msg":"Getting profile for 10-0-28-163.example-cluster.example.svc.cluster.local:8080","remote":{"IP":"10.0.27.105","Port":35686,"Zone":""},"time":"2024-05-24T12:29:12Z"}
{"addr":":8086","component":"workload-watcher","level":"debug","msg":"Establishing watch on workload 10-0-28-163.example-cluster.example:8080","time":"2024-05-24T12:29:12Z"}

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x18bfde1]
```

The fact of using Deployments with headless services is the key in this case because pods are addressable with `pod-ip.service.namespace.svc.cluster.domain`. When the destination controller gets a watch request of this form, they attempt to find a pod with hostname of pod-ip and the `getEndpointByHostname` function in the `WorkloadWatcher` struct was trying to get the endpoint by hostname, but the hostname was null with a nil pointer reference in [this point](https://github.com/linkerd/linkerd2/blob/main/controller/api/destination/watcher/workload_watcher.go#L583):


```go
if hostname == *ep.Hostname
```

### How can it be reproduced?

Establishing a watch on a workload with a headless service and a deployment using the edge-24.3.4 version of Linkerd when the hostname is the pod-ip.

### Logs, error output, etc

The complete error log is as follow is the debug level enabled

```log
{"addr":":8086","component":"server","context-ns":"example-namespace","context-pod":"example-pod-6b686c4df-fh4km","level":"debug","msg":"Getting profile for 10-0-28-163.example-cluster.example.svc.cluster.local:8080","remote":{"IP":"10.0.27.105","Port":35686,"Zone":""},"time":"2024-05-24T12:29:12Z"}
{"addr":":8086","component":"workload-watcher","level":"debug","msg":"Establishing watch on workload 10-0-28-163.example-cluster.example:8080","time":"2024-05-24T12:29:12Z"}

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x18bfde1]

goroutine 1375 [running]:
github.com/linkerd/linkerd2/controller/api/destination/watcher.(*WorkloadWatcher).getEndpointByHostname(0xc0003fa6e0, {0xc003c29040, 0xb}, 0xc004395508)
        /linkerd-build/controller/api/destination/watcher/workload_watcher.go:583 +0x381
github.com/linkerd/linkerd2/controller/api/destination/watcher.(*WorkloadWatcher).getOrNewWorkloadPublisher(0xc0003fa6e0, 0xc004395508, {0xc003c29040, 0xb}, {0x0, 0x0}, 0x1f90)
        /linkerd-build/controller/api/destination/watcher/workload_watcher.go:431 +0xf0
github.com/linkerd/linkerd2/controller/api/destination/watcher.(*WorkloadWatcher).Subscribe(0xc0003fa6e0, 0xc004395508, {0xc003c29040, 0xb}, {0x0, 0x0}, 0x1f90, {0x23b5a60, 0xc0040de380})
        /linkerd-build/controller/api/destination/watcher/workload_watcher.go:120 +0x2e8
github.com/linkerd/linkerd2/controller/api/destination.(*server).subscribeToEndpointProfile(0xc000a01290, 0xc004395508, {0xc003c29040, 0xb}, {0x0, 0x0}, 0x1f90, 0xc0040de2a0, {0x23dea70, 0xc002d7f140})
        /linkerd-build/controller/api/destination/server.go:520 +0x195
github.com/linkerd/linkerd2/controller/api/destination.(*server).getProfileByName(0xc000a01290, {{0xc003c01070, 0xa}, {0xc003c1e980, 0x1b}, {0xc003c1e9a0, 0x1a}}, {0xc003c29040, 0x3b}, 0x1f90, ...)
        /linkerd-build/controller/api/destination/server.go:343 +0xee
github.com/linkerd/linkerd2/controller/api/destination.(*server).GetProfile(0xc000a01290, 0xc0040c9e60, {0x23dea70, 0xc002d7f140})
        /linkerd-build/controller/api/destination/server.go:293 +0x96c
github.com/linkerd/linkerd2-proxy-api/go/destination._Destination_GetProfile_Handler({0x1f59da0, 0xc000a01290}, {0x23dcae8, 0xc002bf7530})
        /go/pkg/mod/github.com/linkerd/linkerd2-proxy-api@v0.13.0/go/destination/destination_grpc.pb.go:172 +0x107
github.com/grpc-ecosystem/go-grpc-prometheus.init.(*ServerMetrics).StreamServerInterceptor.func4({0x1f59da0, 0xc000a01290}, {0x23dc668, 0xc0025df0e0}, 0xc002bf7518, 0x21c27c0)
        /go/pkg/mod/github.com/grpc-ecosystem/go-grpc-prometheus@v1.2.0/server_metrics.go:121 +0xd2
google.golang.org/grpc.(*Server).processStreamingRPC(0xc0001a5a00, {0x23d6768, 0xc0040d4e10}, {0x23e0ee0, 0xc00334c820}, 0xc0040d1560, 0xc000a24f60, 0x34d53e0, 0x0)
        /go/pkg/mod/google.golang.org/grpc@v1.62.1/server.go:1687 +0x11e7
google.golang.org/grpc.(*Server).handleStream(0xc0001a5a00, {0x23e0ee0, 0xc00334c820}, 0xc0040d1560)
        /go/pkg/mod/google.golang.org/grpc@v1.62.1/server.go:1801 +0xe36
google.golang.org/grpc.(*Server).serveStreams.func2.1()
        /go/pkg/mod/google.golang.org/grpc@v1.62.1/server.go:1027 +0x8b
created by google.golang.org/grpc.(*Server).serveStreams.func2 in goroutine 1374
        /go/pkg/mod/google.golang.org/grpc@v1.62.1/server.go:1038 +0x125
```

### output of `linkerd check -o short`

```bash
linkerd-version
---------------
‼ cli is up-to-date
    is running version 24.3.4 but the latest edge version is 24.5.5
    see https://linkerd.io/2/checks/#l5d-version-cli for hints

control-plane-version
---------------------
‼ control plane is up-to-date
    is running version 24.3.4 but the latest edge version is 24.5.5
    see https://linkerd.io/2/checks/#l5d-version-control for hints

linkerd-control-plane-proxy
---------------------------
‼ control plane proxies are up-to-date
    some proxies are not running the current version:
        * linkerd-destination-d954b8596-5fqph (edge-24.3.4)
        * linkerd-destination-d954b8596-6gdrm (edge-24.3.4)
        * linkerd-destination-d954b8596-79s2l (edge-24.3.4)
        * linkerd-identity-769fcf8d4b-2mgx2 (edge-24.3.4)
        * linkerd-identity-769fcf8d4b-lcjxc (edge-24.3.4)
        * linkerd-identity-769fcf8d4b-tj8lr (edge-24.3.4)
        * linkerd-proxy-injector-5dc7466c78-46s79 (edge-24.3.4)
        * linkerd-proxy-injector-5dc7466c78-gdhgm (edge-24.3.4)
        * linkerd-proxy-injector-5dc7466c78-m9gmg (edge-24.3.4)
    see https://linkerd.io/2/checks/#l5d-cp-proxy-version for hints

linkerd-jaeger
--------------
‼ jaeger extension proxies are up-to-date
    some proxies are not running the current version:
        * collector-79c44b64db-rg5ws (edge-24.3.4)
        * jaeger-injector-56556bdf65-ch5s8 (edge-24.3.4)
    see https://linkerd.io/2/checks/#l5d-jaeger-proxy-cp-version for hints

linkerd-multicluster
--------------------
‼ multicluster extension proxies are up-to-date
    some proxies are not running the current version:
        * linkerd-gateway-647876d5b8-9tkdg (edge-24.3.4)
        * linkerd-gateway-647876d5b8-j5wkq (edge-24.3.4)
        * linkerd-service-mirror-monitoring-testing-5c4bbd66bd-97qpk (edge-24.3.4)
    see https://linkerd.io/2/checks/#l5d-multicluster-proxy-cp-version for hints

linkerd-viz
-----------
‼ viz extension proxies are up-to-date
    some proxies are not running the current version:
        * metrics-api-598599fd5f-8cnms (edge-24.3.4)
        * metrics-api-598599fd5f-rncb7 (edge-24.3.4)
        * tap-5b954bb5c-dsgkh (edge-24.3.4)
        * tap-5b954bb5c-t2nwd (edge-24.3.4)
        * tap-injector-69ff64bdc-54f9f (edge-24.3.4)
        * tap-injector-69ff64bdc-8mpq2 (edge-24.3.4)
        * web-66869c8dcc-gnqwn (edge-24.3.4)
        * web-66869c8dcc-rgfzc (edge-24.3.4)
    see https://linkerd.io/2/checks/#l5d-viz-proxy-cp-version for hints
```

### Environment

- Kubernetes version: `1.27`
- Cluster environment: `EKS`
- Host OS: `Linux`
- Linkerd version: `edge-24.3.4`

### Possible solution

A fix to the issue is achieve by checking if the hostname was null before calling the above in the `getEndpointByHostname` function:

```go
if ep.Hostname != nil && hostname == *ep.Hostname
```

After testing it using a forked image for the destination container and using and app with a deployment and a headless service while running a simulation with puppeteer and linkerd injected, the panic error was not triggered, and the simulation ran without causing the linkerd destination container to crash.

Furthermore, I included the next logs to show that the hostname was null in the EndpointSlice:

```go
ww.log.Debugf("No hostname found in EndpointSlice %s/%s for hostname %s with endpoint %v", svcID.Namespace, svcID.Name, hostname, ep)
```

The logs I have obtained when running the simulation with the fixed image are the following:

```log
{"addr":":8086","component":"workload-watcher","level":"debug","msg":"No hostname found in EndpointSlice example-namespace/example-cluster for hostname 10-0-30-175 with endpoint {[10.0.30.175] {0xc0018fb06f 0xc0018fb070 0xc0018fb071} \u003cnil\u003e \u0026ObjectReference{Kind:Pod,Namespace:example-namespace,Name:example-pod-85687854c-566cr,UID:b47831ec-92ef-4d7d-95bd-ea1836b89354,APIVersion:,ResourceVersion:,FieldPath:,} map[] 0xc0018f3d20 0xc0018f3d30 nil}","time":"2024-06-04T16:00:35Z"}
```

As we can see there the `nil` value in the log entry corresponds to the `Hostname` field of the endpoint (`ep`) in the `EndpointSlice`. This field is `nil` because the `EndpointSlice` does not have a hostname assigned for this endpoint. The `Hostname` field in the `EndpointSlice` can be `nil` because headless services do not require DNS names for individual pods, which is the case when using a headless service with a deployment.

If we break down the log entry, we can see the following details:

- **`[10.0.30.175]`**: The IP address of the endpoint.
- **`{0xc0018fb06f 0xc0018fb070 0xc0018fb071}`**: Internal details of the endpoint (likely representing conditions or readiness states).
- **`\u003cnil\u003e`**: This is the `nil` value for the `Hostname` field, encoded in JSON as `\u003cnil\u003e`.
- **`&ObjectReference{Kind:Pod,Namespace:painkiller,Name:painkiller-85687854c-566cr,UID:b47831ec-92ef-4d7d-95bd-ea1836b89354,APIVersion:,ResourceVersion:,FieldPath:,}`**: The `TargetRef` field, which references the pod associated with this endpoint.
- **`map[]`**: An empty map, possibly representing additional metadata or labels.
- **`0xc0018f3d20 0xc0018f3d30 nil`**: Additional internal details of the endpoint.


### Additional context

After applying that bug fix it fails to subscribe to the profile using the pod-ip, but I think that is expected:

```log
{"addr":":8086","component":"server","context-ns":"example-namespace","context-pod":"example-pod-85687854c-sftd6","level":"error","msg":"Failed to subscribe to profile by name \"10-0-27-125.example-cluster.example.svc.cluster.local:8080\": \"rpc error: code = NotFound desc = no pod found in EndpointSlices of Service example-namespace/example-cluster for hostname 10-0-27-125\"","remote":{"IP":"10.0.30.122","Port":51426,"Zone":""},"time":"2024-06-04T16:00:36Z"}
```

### Workaround

A potential workaround for this issue is to use a StatefulSet instead of a Deployment. StatefulSets create stable network identities and the DNS names use the canonical name of the form `hostname.service.namespace.svc.cluster.domain`, not the pod IP. Furthermore, EndpointSlices of headless services deployed with Deployments have the hostname set to null which is not the case when using StatefulSets. This change could prevent the panic error in the Linkerd destination container but the bug will still present.

### Would you like to work on fixing this bug?

yes

--- a/controller/api/destination/test_util.go
+++ b/controller/api/destination/test_util.go
@@ -358,6 +358,8 @@ metadata:
 addressType: IPv4
 endpoints:
 - addresses:
+  - 172.17.13.14 # Endpoint without a targetRef or hostname
+- addresses:
   - 172.17.13.15
   hostname: pod-0
   targetRef:

--- a/controller/api/destination/watcher/workload_watcher.go
+++ b/controller/api/destination/watcher/workload_watcher.go
@@ -580,7 +580,7 @@ func (ww *WorkloadWatcher) getEndpointByHostname(hostname string, svcID *Service
 		}
 		for _, slice := range sliceList {
 			for _, ep := range slice.Endpoints {
-				if hostname == *ep.Hostname {
+				if ep.Hostname != nil && hostname == *ep.Hostname {
 					if ep.TargetRef != nil && ep.TargetRef.Kind == "Pod" {
 						podName := ep.TargetRef.Name
 						podNamespace := ep.TargetRef.Namespace

